tests/schema/tests.py

```diff
@@ -2758,6 +2758,53 @@ def test_alter_field_default_doesnt_perform_queries(self):
         with connection.schema_editor() as editor, self.assertNumQueries(0):
             editor.alter_field(AuthorWithDefaultHeight, old_field, new_field, strict=True)
 
+    @skipUnlessDBFeature('supports_foreign_keys')
+    def test_alter_field_fk_attributes_noop(self):
+        """
+        No queries are performed when changing FK field attributes that
+        don't affect the schema.
+        """
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+            editor.create_model(Book)
+        old_field = Book._meta.get_field('author')
+        new_field = ForeignKey(
+            Author, blank=True, editable=False, error_messages={'invalid': 'new error message'},
+            help_text='new help text', limit_choices_to={'foo': 'bar'}, related_name='new_related_name',
+            related_query_name='new_related_query_name', on_delete=PROTECT, validators=[lambda x: x],
+            verbose_name='new verbose name', db_column=old_field.column
+        )
+        new_field.set_attributes_from_name('renamed_author')
+        with connection.schema_editor() as editor, self.assertNumQueries(0):
+            editor.alter_field(Book, old_field, new_field, strict=True)
+
+    @skipUnlessDBFeature('supports_foreign_keys')
+    def test_alter_field_m2m_attributes_noop(self):
+        """
+        No queries are performed when changing M2M field attributes that
+        don't affect the schema.
+        """
+        class BookWithTags(Model):
+            title = CharField(max_length=100, db_index=True)
+            tags = ManyToManyField(TagM2MTest, related_name='books')
+
+            class Meta:
+                apps = new_apps
+                app_label = 'schema'
+
+        with connection.schema_editor() as editor:
+            editor.create_model(TagM2MTest)
+            editor.create_model(BookWithTags)
+        old_field = BookWithTags._meta.get_field('tags')
+        new_field = ManyToManyField(
+            TagM2MTest, blank=True, error_messages={'invalid': 'new error message'}, help_text='new help text',
+            limit_choices_to={'foo': 'bar'}, related_name='new_related_name',
+            related_query_name='new_related_query_name', verbose_name='new verbose name',
+        )
+        new_field.set_attributes_from_name('tags')
+        with connection.schema_editor() as editor, self.assertNumQueries(0):
+            editor.alter_field(BookWithTags, old_field, new_field, strict=True)
```

Comment:
Altering m2m fields in tests is really tricky, because we should call `contribute_to_class()` on a new field to set all necessary attributes and methods. We cannot do this with the same name because it will raise a warning that an intermediate model is already registered. We cannot also do this with a different name because it will cause renaming of an intermediate model and `assertNumQueries(0)` will fail. So without the patch `test_alter_field_m2m_attributes_noop` will crash with:

```
ValueError: Cannot alter field schema.BookWithTags.tags into <django.db.models.fields.related.ManyToManyField: tags> - they are not compatible types (you cannot alter to or from M2M fields, or add or remove through= on M2M fields)
```

I think it's fine to chop this test because it's not testing a different path, there is nothing m2m in the fix.